syntax = "proto2";

message DroneMessage {
	enum FlightMode {
		m_degraded = 0;
		m_passThrough = 1;
		m_flybywire = 2;
		m_autonomous = 3;
		m_landed = 4;
		mx_autonomous_ai = 5;
		mx_unavailable = 6;
	}

	message Position {
		optional int32 timestamp = 20;
		optional int32 real_time = 21;
	
		required sint32 latitude = 22;
		required sint32 longitude = 23;
		optional int32 altitude = 24;
	}
	
	message Velocity {	
		optional int32 timestamp = 30;
	
		optional int32 speed = 31;
		optional int32 course = 32;
		optional sint32 rate_of_climb = 33;
	}
	
	message Attitude {
		optional int32 timestamp = 40;
		
		required sint32 yaw = 41;
		required sint32 pitch = 42;
		required sint32 roll = 43;
	}
	
	message BaroData {
		required int32 timestamp = 50;
		
		required float pressure = 51;
		required float temperature = 52;
	}
	
	message ThreeDimensionalData {
		required int32 timestamp = 60;
		
		required float x = 62;
		required float y = 63;
		required float z = 64;
	}
	
	optional FlightMode current_mode = 1;
	optional Velocity current_speed = 2;
	optional Velocity current_position = 3;
	optional Attitude current_attitude = 4;
	
	optional BaroData bmp_raw = 5;
	optional ThreeDimensionalData gyro_raw = 6;
	optional ThreeDimensionalData mag_raw = 7;
	optional ThreeDimensionalData accel_raw = 8;
	
	optional int32 battery_current = 9;
	optional int32 battery_voltage = 10;
	optional int32 battery_percentage = 11;
	
	//Now the "input" messages
	optional Position new_waypoint = 12;
	optional Position new_home_base = 13;
	optional float sea_level_pressure = 14;
	optional Attitude command_set = 15;
	optional FlightMode new_mode = 16;
}